src/frontend/Semantic_check.ml

```diff
@@ -347,6 +348,92 @@ let semantic_check_reduce_sum ~is_cond_dist ~loc id es =
       |> Semantic_error.illtyped_reduce_sum_generic loc id.name
       |> Validate.error
 
+let semantic_check_variadic_ode_tol ~is_cond_dist ~loc id es =
+  let arg_match (x_ad, x_t) y =
+    UnsizedType.check_of_same_type_mod_conv "" x_t y.emeta.type_
+    && UnsizedType.autodifftype_can_convert x_ad y.emeta.ad_level
+  in
+  let args_match a b =
+    List.length a = List.length b && List.for_all2_exn ~f:arg_match a b
+  in
+  match es with
+  | { emeta=
+        { type_=
+            UnsizedType.UFun
+              ( (_, UReal) (* time *)
+                
+                :: (_, UnsizedType.UVector) (* initial_state *)
+                    :: fun_args
+              , ReturnType UnsizedType.UVector ); _ }; _ }
+    :: initial_state
+       :: initial_time
+          :: times
+             :: rel_tol
+                :: abs_tol
+                   :: { emeta= {type_= UInt; ad_level= UnsizedType.DataOnly; _}; _
+                      } (* max_num_steps *)
+                      
+                      :: args
+    when arg_match (AutoDiffable, UReal) initial_time
+         && arg_match (AutoDiffable, UVector) initial_state
+         && arg_match (AutoDiffable, UArray UReal) times
+         && arg_match (DataOnly, UReal) rel_tol
+         && arg_match (DataOnly, UReal) abs_tol ->
```

Comment:
It seems like these two lines and error types are the only difference between this function and semantic_check_variadic_ode? Can we call that one here, or have them share a helper function? That'd save ~50 lines of replicated code